### PR TITLE
[types-2.0] react - better strict mode compatibility

### PR DESCRIPTION
**Pull request by beckend**
# index.d.ts
- line 16, `Key` was changed to accept null, When you do `React.createElement(...)` and you do not supply a key, it becomes the value `null`.
- line 40, `ComponentClass<P>`, when testing `React.cloneElement` which is able to accept `null` and `undefined`, had to add `any` type.
- line 77, `DOMFactory<P, T>`, using any `React.DOM.xxx` function to create an element. For example `React.DOM.div(null, undefined)` actually works and thus added.
- line 96, a React child can be `null`.
- line 202, the last [commit](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/b329e7ec15c7a6dbebfca0e0ed916cb5f60f269a) added the return type of `null` to `SFC<P>` and in strict mode there was no longer the possibility to render the component any more, it will say `null is not a constructor function for jsx elements.`. A workaround is to convert it to an interface then overload the constructor function to have let `SFC<P>` to have the ability to return null, but leave `StatelessComponent<P>` alone. You now have 2 versions to work with, in strict mode `SFC<P>` will fail to render every time because of `null` return, but `StatelessComponent<P>` will work.
- line 265, `ComponentSpec.render`, the render may return `null` like every other render.

---

The rest of the files had to be adjusted accordingly when using `strictNullChecks`.

### CODE REVIEW

**Comment by beckend** [comment]
I don't think that this covers all `null` or `undefined` clauses but it's a start.

**Comment by donaldpipowitch** [comment]
`SFC` was an alias to `StatelessComponent`. I'd like to keep this behaviour. I would _not_ want to decide which interface I should use depending on my current tsconfig. For now I'd just want to revert my change: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11931.

I think this could be a problem with TypeScript itself, because it only happens with TSX and not with TS. I reported this here: https://github.com/Microsoft/TypeScript/issues/11566.

**Comment by beckend** [comment]
Let me know how you want to proceed.

**Comment by donaldpipowitch** [comment]
You could include my PR #11931 and I'll close mine. The bug is reported now and I'll add `| null` back to the definitions when it works with TSX.

**Comment by beckend** [comment]
It is done.

**Comment by donaldpipowitch** [comment]
Cool. I've closed my PR.

// cc @aozgaa You merged my previous PR. Could you review this PR? Thanks.

**Comment by donaldpipowitch** [comment]
@vvakame Do you know why this PR fails here?:

```
----------------- For file:/home/travis/build/DefinitelyTyped/DefinitelyTyped/ejs/tsconfig.json
\home\travis\build\DefinitelyTyped\DefinitelyTyped\ejs\index.d.ts
	 Line 11: Cannot redeclare exported variable 'cache'.
\home\travis\build\DefinitelyTyped\DefinitelyTyped\ejs\index.d.ts
	 Line 65: Cannot redeclare exported variable 'cache'.
```

I doesn't look like this PR touches `ejs`?

**Comment by vvakame** [comment]
umm... Could you try to merge types-2.0/HEAD to your branch?

**Comment by beckend** [comment]
@vvakame good call, that fixed it.

**Comment by vvakame** [comment]
*react/index.d.ts*

to authors (@pspeter3 @vsiao @johnnyreilly AssureSign (account can't be detected) Microsoft (account can't be detected)). Could you review this PR?
:+1: or :-1:?

Checklist

* [ ] pass the Travis CI test?

**Comment by vvakame** [comment]
When we can't get response from author. please ping me after 3 days. I'll merge this PR.

File: react/index.d.ts [hunk 1]
```diff
@@ -74,7 +74,7 @@ declare namespace React {
     type ClassicFactory<P> = CFactory<P, ClassicComponent<P, ComponentState>>;
 
     interface DOMFactory<P extends DOMAttributes<T>, T extends Element> {
-        (props?: P & ClassAttributes<T>, ...children: ReactNode[]): DOMElement<P, T>;
+        (props?: P & ClassAttributes<T> | null, ...children: Array<ReactNode | undefined | null>): DOMElement<P, T>;
```
Comment (vsiao) [inline react/index.d.ts]: `children` should remain `ReactNode[]` and `ReactNode` should be fixed to allow `undefined` and `null`.

File: react/index.d.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ declare namespace React {
 
     type CElement<P, T extends Component<P, ComponentState>> = ComponentElement<P, T>;
     interface ComponentElement<P, T extends Component<P, ComponentState>> extends ReactElement<P> {
-        type: ComponentClass<P>;
+        type: ComponentClass<P> | ComponentClass<any>;
```
Comment (vsiao) [inline react/index.d.ts]: What is this for?

File: react/react-tests.ts [hunk 3]
```diff
@@ -41,19 +41,20 @@ var props: Props & React.ClassAttributes<{}> = {
     foo: 42
 };
 
-var container: Element;
+var container: Element = document.createElement('div');
```
Comment (vsiao) [inline react/react-tests.ts]: double quotes please

File: react/react-tests.ts [hunk 4]
```diff
@@ -145,7 +146,7 @@ namespace StatelessComponent {
 
 var StatelessComponent2: React.SFC<SCProps> =
     // props is contextually typed
-    props => React.DOM.div(null, props.foo);
+    (props: SCProps) => React.DOM.div(null, props.foo);
```
Comment (vsiao) [inline react/react-tests.ts]: per the comment above this line, `props` is supposed to be contextually typed. So please leave out the `SCProps` annotation.

File: react/react-tests.ts [hunk 5]
```diff
@@ -240,18 +247,15 @@ domNode = ReactDOM.findDOMNode(domNode);
...
 
 //
 // React Components
 // --------------------------------------------------------------------------
 
-var displayName: string = ClassicComponent.displayName;
-var defaultProps: Props = ClassicComponent.getDefaultProps();
-var propTypes: React.ValidationMap<Props> = ClassicComponent.propTypes;
+var displayName: string | undefined = ClassicComponent.displayName;
+var defaultProps: Props | {} = ClassicComponent.getDefaultProps ? ClassicComponent.getDefaultProps() : {};
```
Comment (vsiao) [inline react/react-tests.ts]: can you keep `: Props =` and cast the empty object (`<Props>{}`) instead? Otherwise, if `getDefaultProps()` has the wrong return type it'll get assigned to `{}` and won't break.

File: react/react-tests.ts [hunk 6]
```diff
@@ -367,14 +371,14 @@ var PropTypesSpecification: React.ComponentSpec<any, any> = {
         }),
         requiredFunc: React.PropTypes.func.isRequired,
         requiredAny: React.PropTypes.any.isRequired,
-        customProp: function(props: any, propName: string, componentName: string) {
+        customProp: function(props: any, propName: string, componentName: string): any {
```
Comment (vsiao) [inline react/react-tests.ts]: `: Error | null`?

File: react/index.d.ts [hunk 7]
```diff
@@ -14,7 +14,7 @@ declare namespace React {
 
     type ReactType = string | ComponentClass<any> | StatelessComponent<any>;
 
-    type Key = string | number;
+    type Key = string | number | null;
```
Comment (vsiao) [inline react/index.d.ts]: Why is this necessary?

**Review by vsiao — CHANGES_REQUESTED** [review]
see above comments

File: react/index.d.ts (hunk 7 above)
Comment (beckend) [inline react/index.d.ts]: You can use `  React.createElement`   or other methods that created elements, the return type should reflect the actual possible types a variable can actually become.

So if you do something with the returned payload it warns you that the value can actually be `  null`  . I don't actually get the question here since the `  Key`   value can actually be null from the built in methods, regardless of necessity, this reflects the actual React API workings.

File: react/index.d.ts (hunk 2 above)
Comment (beckend) [inline react/index.d.ts]: This is when you use for example `  React.cloneElement`  .
The elements can have `  ComponentClass<undefined>`   or `  ComponentClass<null>`   which will error.
Should you revert this change you can't clone elements in strictNullChecks or other methods that accepts `  ComponentClass`   with either null or undefined.

This is of course can be simplified by only use `  any`   but I wanted to leave some traces of what is was aiming to be to.

File: react/index.d.ts (hunk 1 above)
Comment (beckend) [inline react/index.d.ts]: Good catch, that will be fixed.

File: react/index.d.ts (hunk 7 above)
Comment (vsiao) [inline react/index.d.ts]: > the `Key` value can actually be null from the built in methods

Great, that's what I'm asking about. Did you consider changing the `key` property on `interface ReactElement<P>` to `key: Key | null` instead?

File: react/react-tests.ts (hunk 3 above)
Comment (beckend) [inline react/react-tests.ts]: Will be fixed.

File: react/index.d.ts (hunk 2 above)
Comment (vsiao) [inline react/index.d.ts]: Can you give an example? Other top-level API methods rely on this being strongly typed (eg. to infer the correct return type of `ReactDOM.render`) so I'm worried that the `any` will break that.

File: react/index.d.ts (hunk 7 above)
Comment (beckend) [inline react/index.d.ts]: That's actually better, see it fixed for next commit.

File: react/react-tests.ts (hunk 4 above)
Comment (beckend) [inline react/react-tests.ts]: Will be fixed.

File: react/react-tests.ts (hunk 6 above)
Comment (beckend) [inline react/react-tests.ts]: Type `  null`   is not assignable to `  Error`  .

File: react/react-tests.ts (hunk 5 above)
Comment (beckend) [inline react/react-tests.ts]: Will be fixed.

**Comment by beckend** [comment]
@vsiao current reviews committed.

File: react/index.d.ts (hunk 2 above)
Comment (beckend) [inline react/index.d.ts]: Yea this is kind of sensitive, but that's what code reviews are for, I'm kind of unsure how to best fix this.

screenshot from 2016-10-30 15-57-08

Comment (beckend) [inline react/index.d.ts]: So the test would become
```js
React.cloneElement(element, {});
React.cloneElement(element, {}, null);
React.cloneElement(element, {});
React.cloneElement(element, {}, null);
```
And pass.

Comment (beckend) [inline react/index.d.ts]: @vsiao can you review this so the react pulls can proceed?

Comment (vsiao) [inline react/index.d.ts]: Sounds like the right way to fix this would be to add an override to `cloneElement` that takes a single `element` argument with no `props` so that the compiler doesn't infer `undefined` for `Props`.

Comment (vsiao) [inline react/index.d.ts]: I suppose that doesn't fix your test case of `React.cloneElement(element, {}, null)`. Is it possible to use a `null` literal type? ie.

```ts
function cloneElement<P>(element: ReactElement<P>, props: null, ...children: ReactNode[]): ReactElement<P>;
```

File: react/react-tests.ts (hunk 6 above)
Comment (vsiao) [inline react/react-tests.ts]: Then the `ValidationMap` interface needs to be fixed, perhaps?

**Comment by yuit** [comment]
@vsiao  is the PR look good to you?

**Comment by vsiao** [comment]
No, not yet. Thanks for asking.

**Comment by beckend** [comment]
@vsiao updated.

File: react/index.d.ts [hunk 8]
```diff
@@ -28,7 +28,7 @@ declare namespace React {
     interface ReactElement<P> {
         type: string | ComponentClass<P> | SFC<P>;
         props: P;
-        key?: Key;
+        key?: Key | null;
```
Comment (vsiao) [inline react/index.d.ts]: Is it possible for `key` to be undefined? If not, let's remove `?`.

File: react/react-tests.ts [hunk 9]
```diff
@@ -41,19 +41,20 @@ var props: Props & React.ClassAttributes<{}> = {
...
-var container: Element;
+var container: Element = document.createElement("div");
 
 //
 // Top-Level API
 // --------------------------------------------------------------------------
 
 var ClassicComponent: React.ClassicComponentClass<Props> =
     React.createClass<Props, State>({
+        displayName: 'ClassicComponent',
```
Comment (vsiao) [inline react/react-tests.ts]: double quotes please

File: react/react-tests.ts [hunk 10]
```diff
@@ -145,7 +146,7 @@ namespace StatelessComponent {
 
 var StatelessComponent2: React.SFC<SCProps> =
     // props is contextually typed
-    props => React.DOM.div(null, props.foo);
+    (props) => React.DOM.div(null, props.foo);
```
Comment (vsiao) [inline react/react-tests.ts]: unrelated change here

File: react/react-tests.ts [hunk 11]
```diff
@@ -367,14 +369,14 @@ var PropTypesSpecification: React.ComponentSpec<any, any> = {
...
-        customProp: function(props: any, propName: string, componentName: string) {
+        customProp: function(props: any, propName: string, componentName: string): Error | null {
             if (!/matchme/.test(props[propName])) {
                 return new Error("Validation failed!");
             }
             return null;
         },
         // https://facebook.github.io/react/warnings/dont-call-proptypes.html#fixing-the-false-positive-in-third-party-proptypes
-        percentage: (object: any, key: string, componentName: string, ...rest: any[]): Error => {
+        percentage: (object: any, key: string, componentName: string, ...rest: any[]): any => {
```
Comment (vsiao) [inline react/react-tests.ts]: `Error | null`

**Review by vsiao — COMMENTED** [review]
@beckend: almost there! just a few nits inline.

File: react/index.d.ts (hunk 8 above)
Comment (beckend) [inline react/index.d.ts]: It's possible.

**Comment by beckend** [comment]
@vsiao bump.

**Comment by donaldpipowitch** [comment]
@beckend  Thank you for your fast responses.

**Comment by beckend** [comment]
I am waiting for this PR as well :)

**Comment by johnnyreilly** [comment]
Is everyone happy for this to be merged now?

**Comment by vsiao** [comment]
lgtm! Thanks for your work, @beckend.

**Comment by donaldpipowitch** [comment]
Hooray!
